### PR TITLE
Upload data app

### PR DESCRIPTION
**Pull request by ben18785**
- adds capacity to upload data
- checks that a number of columns in Roche data are present
- if missing values are encountered the user is warned that they'll be dropped then, if willing to accept this, they can proceed by re-uploading (this is a placeholder for what we will do later on when we use Roche's missing data system)
- checks that certain time units (h, d, hours) are present for unit conversion
- assumes rows where `DV=0` are doses

### CODE REVIEW

**Comment by martinjrobins** [comment]
thanks @ben18785 , do you have a test dataset that works for the upload?

File: pkpdapp/pkpdapp/forms/dataset.py [hunk 1]
```diff
@@ -15,58 +20,219 @@ def file_size(value):
...
+        for index, row in bts_unique.iterrows():
+            if row['UNIT'] == ".":
+                BiomarkerType.objects.create(
+                    name=row['YDESC'],
+                    description="",
+                    dataset=instance)
+            else:
+                unit_query = StandardUnit.objects.filter(symbol=row['UNIT'])
+                if not unit_query:
+                    unit = StandardUnit(symbol=row['UNIT'])
```
Comment (martinjrobins) [inline pkpdapp/pkpdapp/forms/dataset.py:156]: I don't think we should add to the list of standard units here, as the user could be uploading any any old units. We should restrict possible units in uploaded datafiles to those in the `Unit` table, and convert these to units in the `StandardUnit` table, as is done in the migrations. So all data is stored in standardised units that we control, and can deal with properly when plotting etc.

File: pkpdapp/pkpdapp/forms/dataset.py [hunk 2]
```diff
@@ -15,58 +20,219 @@ def file_size(value):
...
-    ]
-    unit = forms.ChoiceField(label='', choices=UNIT_CHOICES)
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    class Meta:
+        model = BiomarkerType
+        fields = ['description']
+
+    symbol = forms.CharField(label='Unit', required=False)
```
Comment (martinjrobins) [inline pkpdapp/pkpdapp/forms/dataset.py:233]: Probably want a `django.ModelChoiceField` here for `Unit`

File: pkpdapp/pkpdapp/migrations/0005_auto_20210318_1408.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
+# Generated by Django 3.0.7 on 2021-03-18 14:08
+
+from django.db import migrations, models
+
+
+class Migration(migrations.Migration):
+
```
Comment (martinjrobins) [inline pkpdapp/pkpdapp/migrations/0005_auto_20210318_1408.py:7]: please have all schema changes merged into the initial 0001 migration

File: pkpdapp/pkpdapp/migrations/0006_auto_20210324_1534.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+from django.db import migrations, models
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('pkpdapp', '0005_auto_20210318_1408'),
+    ]
+
+    operations = [
```
Comment (martinjrobins) [inline pkpdapp/pkpdapp/migrations/0006_auto_20210324_1534.py:12]: this one too

File: pkpdapp/pkpdapp/templates/base/head.html [hunk 5]
```diff
@@ -46,8 +46,8 @@
                 <a class="dropdown-item" href={% url 'dosed_pk_model-detail' m.id
                   %}>{{m.pharmacokinetic_model.name}}</a>
                 {% endfor %}
-                <a class="dropdown-item" href="{% url 'dosed_pk_model-add-to-project' current_project.id %}"> 
-                  Create model.. 
+                <a class="dropdown-item" href="{% url 'dosed_pk_model-add-to-project' current_project.id %}">
+                  Create model..
```
Comment (martinjrobins) [inline pkpdapp/pkpdapp/templates/base/head.html:50]: Can you please add a similar "Add dataset" action link for the dataset dropdown list?

File: pkpdapp/pkpdapp/templates/dataset_form.html [hunk 6]
```diff
@@ -11,7 +11,7 @@
 
 <h2>Update Dataset {{dataset.name}}</h2>
 
-<form action="" method="post">
+<form action="" method="post"  enctype="multipart/form-data">
   {% csrf_token %}
```
Comment (martinjrobins) [inline pkpdapp/pkpdapp/templates/dataset_form.html:15]: Can we please add some text to this page detailing the csv format required for upload? i.e. required columns

**Review by martinjrobins — CHANGES_REQUESTED** [review]
Thanks @ben18785, worked great for me. But I think there needs to be unit conversions to make sure that all uploaded data is in the same set of standard units in the `StandardUnit` table, so we can then plot it sensibly or use it for dosing.

File: pkpdapp/pkpdapp/migrations/0005_auto_20210318_1408.py (hunk 3 above)
Comment (ben18785) [reply pkpdapp/pkpdapp/migrations/0005_auto_20210318_1408.py:7]: Done

File: pkpdapp/pkpdapp/migrations/0006_auto_20210324_1534.py (hunk 4 above)
Comment (ben18785) [reply pkpdapp/pkpdapp/migrations/0006_auto_20210324_1534.py:12]: Done

File: pkpdapp/pkpdapp/templates/base/head.html (hunk 5 above)
Comment (ben18785) [reply pkpdapp/pkpdapp/templates/base/head.html:50]: Done

File: pkpdapp/pkpdapp/forms/dataset.py (hunk 1 above)
Comment (ben18785) [reply pkpdapp/pkpdapp/forms/dataset.py:156]: Done.

File: pkpdapp/pkpdapp/forms/dataset.py (hunk 2 above)
Comment (ben18785) [reply pkpdapp/pkpdapp/forms/dataset.py:233]: Done.

**Pull request merged**